### DISCUSSION TITLE
The problem to understand  vprec backend

### DISCUSSION THREAD

**Discussion by Kim1230** [verificarlo/verificarlo#330] [Q&A]
Hello, 
I have two questions about the vprec mode. 
First, I read the src of the vprec backend and get a question about the --max-abs-error-exponent=ERR_EXPONENT and --error-mode=ERR_MODE  select error mode among (rel, abs, all). I can see that this will be used in function `handle_binary32_normal_absErr` and `compute_absErr_vprec_binary64`. The exponent difference is related to the precision adjusted. I don't know why the precision should  be adjusted because of expDiff and the meaning to set this error mode. 
Second, I see vprec will do like this:
![image](https://user-images.githubusercontent.com/107012598/207026852-64195ebc-9eff-42a5-ba9f-f964a20c0061.png)
And I test my application when I change double  variables to float variables or even only export vprec_backend and don't set the precision and range. Compared to the gcc compilation, when I export VPREC_backend and set precision  and range or not, the running time will extend from 1h to 16h. I don't know why the verificarlo costs too much time? which part is the hot pot ? 
Thanks for your help and waitting for your response.

**Comment by pablooliveira** [comment]
Hi @Kim1230.

The functions you pointed are used for vprec absolute mode. In the default relative mode, we simulate a reduced number of precision digits. On the other hand, in absolute mode we always add a constant noise level which is independent of the exponent.
The implementation of the absolute mode was done by @mistoan and @ddavid42.

Regarding the instrumentation cost (which in your application is a x16 overhead) it comes from different effects: instrumenting the code replaces vectorized operations by scalar ones, we also need an indirect call to the verificarlo backend, finally depending on the backend, we need to do additional operations compared to the hardware FP implementation.

**Reply by Kim1230** [reply]
@pablooliveira
Thanks for your response. I want to continue the discussion about the second question. Did you mean the replacement from the vectorized operations to scalar ones will cost too much overhead? How I can check it or find which part cost most? (Where I can use the time function to record the time ?)
I use perf to watch the hot pot. 
![image](https://user-images.githubusercontent.com/107012598/207038025-b11c6304-985c-4208-969b-d695b7f48766.png)
And is there anything to do to reduce the time costs? 
Hope for your response!

**Reply by mistoan** [reply]
I'm not sure you can gain much by changing the mode in vprec.
Most of the cost seems to come from the instrumentation of the code, as your screenshot shows.

You might lower your execution times by focusing the instrumentation on specific parts of the code.
Verificarlo allows doing this through white-lists.

**Reply by pablooliveira** [reply]
Verificarlo breaks down vectorized operations: for example for vectors of width 4, verificarlo will replace one vectorized operation by four scalar operations. Therefore, with verificarlo, vectorized codes run slower.

Regarding optimizations,looking at the code with `perf` is a good idea. To optimize overhead, you can try to rewrite parts of the backend in a more efficient way or remove things that you do not need for your specific study. In any case, if you find something to optimize, feel free to contribute a « Pull Request ».

**Reply by Kim1230** [reply]
Hello@mistoan,
I need your help about question one. The absolute error mode in the vprec. It computes expDiff  and compare it with the preicison and even adjust the precision. Why the exponent will affect the precision and this is only in the absolute error mode not the relative error mode ? Under which circumstance,  we need to consider the absolute error mode ?
Hope for your response. Thanks
